pandas/core/algorithms.py

```diff
@@ -1876,21 +1876,21 @@ def _sort_tuples(values: np.ndarray) -> np.ndarray:
     return values[indexer]
 
 
-def union_with_duplicates(lvals: np.ndarray, rvals: np.ndarray) -> np.ndarray:
+def union_with_duplicates(lvals: ArrayLike, rvals: ArrayLike) -> ArrayLike:
     """
     Extracts the union from lvals and rvals with respect to duplicates and nans in
     both arrays.
 
     Parameters
     ----------
-    lvals: np.ndarray
+    lvals: ArrayLike
         left values which is ordered in front.
-    rvals: np.ndarray
+    rvals: ArrayLike
         right values ordered after lvals.
 
     Returns
     -------
-    np.ndarray containing the unsorted union of both arrays
+    ArrayLike containing the unsorted union of both arrays
```

Comment:
im finding myself getting here on another branch.  i think the pd_array call is causing problems.  In the case I'm looking at, lvals and rvals are both all-string Categoricals, so the pd_array call gives a StringArray, which gums up the works later in Index.union

Reply:
If we say that we should not get here with an extension dtype except categorical, we could remove the pd.array call?

Alternatively if we overwrite the categorical _union, this would be solved too?

Reply:
> Alternatively if we overwrite the categorical _union, this would be solved too?

I think we do this on ExtensionIndex, that way we can do `cast(np.ndarray, lvals)` in `Index._union`.  All ExtensionIndex subclasses _other_ than CategoricalIndex already override _union, so it should be equivalent

Reply:
so i think the following works for CategoricalIndex._union

```
    def _union(self, other: CategoricalIndex, sort) -> CategoricalIndex:
        # we only get here with matching dtypes
        lidx = self.astype(self.categories.dtype)
        ridx = other.astype(other.categories.dtype)
        result = lidx._union(ridx, sort=sort)
        cat = Categorical(result, dtype=self.dtype)
        return type(self)._simple_new(cat, name=self.name)
```

(this might also make it possible to remove a Categorical kludge in `Index._wrap_setop_result`)

The downside here is that the astype both makes a copy and loses potentially-cached is_monotonic/has_duplicates/is_unique information.  It may be better to make a method just for _union_non_unique and only override that.

Reply:
i take it back, that _union messes up when self.dtype.order is non-standard